third_party/souffle.BUILD

```diff
@@ -0,0 +1,465 @@
+        "src/ram/IndexChoice.h",
+        "src/ram/IndexOperation.h",
+        "src/ram/IndexScan.h",
+        "src/ram/IntrinsicOperator.h",
+        "src/ram/ListStatement.h",
+        "src/ram/LogRelationTimer.h",
+        "src/ram/LogSize.h",
+        "src/ram/LogTimer.h",
+        "src/ram/Loop.h",
+        "src/ram/Negation.h",
+        "src/ram/NestedIntrinsicOperator.h",
+        "src/ram/NestedOperation.h",
+        "src/ram/Node.h",
+        "src/ram/Operation.h",
+        "src/ram/PackRecord.h",
+        "src/ram/Parallel.h",
+        "src/ram/ParallelAggregate.h",
+        "src/ram/ParallelChoice.h",
+        "src/ram/ParallelIndexAggregate.h",
+        "src/ram/ParallelIndexChoice.h",
+        "src/ram/ParallelIndexScan.h",
+        "src/ram/Program.h",
+        "src/ram/Project.h",
+        "src/ram/ProvenanceExistenceCheck.h",
+        "src/ram/Query.h",
+        "src/ram/Relation.h",
+        "src/ram/RelationOperation.h",
+        "src/ram/RelationStatement.h",
+        "src/ram/RelationSize.h",
+        "src/ram/ParallelScan.h",
+        "src/ram/Scan.h",
+        "src/ram/Sequence.h",
+        "src/ram/SignedConstant.h",
+        "src/ram/Statement.h",
+        "src/ram/SubroutineArgument.h",
+        "src/ram/SubroutineReturn.h",
+        "src/ram/Swap.h",
+        "src/ram/TranslationUnit.h",
+        "src/ram/True.h",
+        "src/ram/TupleElement.h",
+        "src/ram/TupleOperation.h",
+        "src/ram/UndefValue.h",
+        "src/ram/UnpackRecord.h",
+        "src/ram/UnsignedConstant.h",
+        "src/ram/UserDefinedOperator.h",
+        "src/ram/analysis/Analysis.h",
+        "src/ram/analysis/Complexity.h",
+        "src/ram/analysis/Index.h",
+        "src/ram/analysis/Level.h",
+        "src/ram/analysis/Relation.h",
+        "src/ram/transform/ChoiceConversion.h",
+        "src/ram/transform/CollapseFilters.h",
+        "src/ram/transform/Conditional.h",
+        "src/ram/transform/EliminateDuplicates.h",
+        "src/ram/transform/ExpandFilter.h",
+        "src/ram/transform/HoistAggregate.h",
+        "src/ram/transform/HoistConditions.h",
+        "src/ram/transform/IfConversion.h",
+        "src/ram/transform/IndexedInequality.h",
+        "src/ram/transform/Loop.h",
+        "src/ram/transform/MakeIndex.h",
+        "src/ram/transform/Meta.h",
+        "src/ram/transform/Parallel.h",
+        "src/ram/transform/ReorderConditions.h",
+        "src/ram/transform/ReorderFilterBreak.h",
+        "src/ram/transform/ReportIndex.h",
+        "src/ram/transform/Sequence.h",
+        "src/ram/transform/Transformer.h",
+        "src/ram/transform/TupleId.h",
+        "src/ram/utility/LambdaNodeMapper.h",
+        "src/ram/utility/NodeMapper.h",
+        "src/ram/utility/Utils.h",
+        "src/ram/utility/Visitor.h",
+        "src/reports/DebugReport.h",
+        "src/reports/ErrorReport.h",
+        "src/synthesiser/Synthesiser.h",
+        "src/synthesiser/Relation.h",
+    ],
+    includes = [
+        "src",
```

Comment:
omit this; `includes` are inherited by everything which transitively depends on a particular cc_library. Obnoxiously, if you want to change the include path while compiling the library itself, you need to do `copts = ["-Iexternal/annoying/bazel/path/to/src"]`

Reply:
I need access to these headers from dependent libraries.

Reply:
Which way would you find acceptable? I expect that I will need access to most of the headers (+ transitive dependencies) that main.cpp requires.

Reply:
Man I hate header files. Assuming it actually works, I think moving the internal headers to srcs here and adding something like:

```
cc_library(
  name = "internal",
  hdrs = ["all/the/things"],
  strip_include_prefix = "src",
  include_prefix = "souffle",  # Or "souffle/internal" ?
)
```

so that external clients, in theory, have to use `#include <souffle/synthesiser/Relation.h>` rather than `#include "src/synthesiser/Relation.h"` would be fine?  I'm not sure whether or not you can get away with doing all of that in the same cc_library or not (but I suspect not). If that doesn't work this is fine; I'd like a little lipstick on the pig, but it's not worth losing a hand.

Reply:
I messed with this for a while and eventually got to:
```
cc_library(
    name = "internal_headers",
    hdrs = glob(["src/**/*.h"], exclude = ["src/include/**"]),
    strip_include_prefix = "src",
    include_prefix = "souffle",
)

cc_library(
    name = "external_headers",
    hdrs = glob(["src/include/**/*.h"]),
    strip_include_prefix = "src/include",
    include_prefix = "souffle",
)

...
    copts = [
        "-I$(GENDIR)/external/souffle/_virtual_includes/internal_headers/souffle",
        "-I$(GENDIR)/external/souffle/_virtual_includes/external_headers/souffle",
        "-I$(GENDIR)/external/souffle/src",
    ],
...
```

I don't actually know if the library would be usable even if it builds, which it doesn't, because:
```
external/souffle/src/ast/analysis/Type.cpp:19:10: fatal error: 'ConstraintSystem.h' file not found
#include "ConstraintSystem.h"
```

ConstraintSystem.h *is* in src/ast/analysis, but it's not mapped there by Bazel inside the sandbox.

Reply:
(`external_headers` isn't necessary here because there's a src/include/souffle, but fixing that won't fix the mapping problem with the internal headers.)

Reply:
Okay, it turns out that Bazel is less concerned about multiple rules claiming the same headers as I thought.

Reply:
Unfortunately this still doesn't work. If I depend on souffle from the verifier and include one of the internal headers, it's not able to find other internal headers it references:
```
bazel-out/darwin-fastbuild/bin/external/souffle/_virtual_includes/internal_headers/souffle/ast/Program.h:19:10: fatal error: 'ast/Clause.h' file not found
#include "ast/Clause.h"
```
If I add the nasty copt (`"-I$(GENDIR)/external/souffle/_virtual_includes/internal_headers/souffle",`) to the relevant verifier target it works. This is gross but maybe not as gross as dumping a bunch of souffle-internal headers all over the search path. WDYT? Maybe there's a better solution I'm missing?

Reply:
Yeah, Bazel is generally fine with rules sharing inputs, it gets crabby when they share outputs.

Thanks for fighting with this. At this point, I think polluting the header space is probably the least bad.  None of the header files that I can see are named `config.h` or something equally generic.

Another option would be to use a genrule to transform the include paths used by the headers you want to re-export, but that seems like overkill here (although, I would definitely suggest it if there were something like `config.h` in use). Sorry for all of the song-and-dance.

Reply:
Okay, in the latest commit I've just gone with `includes`.